### PR TITLE
Implement git refs

### LINKED ISSUE #73: Support git refs
Right now fetch supports `--commit`, `--branch`, and `--tag`, but it requires you to know which type you are passing in. This can be limiting especially in CI where the logic to know if the build is on a tag, branch, or commit sha can be quite complex depending on the platform.

Propose a unified `--ref` flag that can be used to pass in any valid git ref and `fetch` does the right thing.

### PR DESCRIPTION
**Pull request by pete0emerson**
This closes #73 

This implements a `--ref` flag which can be used instead of `--tag`, `--branch`, or `--commit`.

### CODE REVIEW

File: README.md
Comment (brikis98) [inline README.md:18]: NIT: perhaps:

```suggestion
- Download from any git reference, such as a specific git tag, branch, or commit SHA.
```

File: README.md
Comment (brikis98) [inline README.md:82]: Consider putting this first. And updating most of the examples to use it, as it's simpler, and probably our preferred option.

File: README.md
Comment (brikis98) [inline README.md:204]: I assume these are just your own notes?

**Review by brikis98 — COMMENTED** [review]
Looks like a good start to me! I made some minor comments on the phrasing, but feel free to use or ignore those, as none are terribly important, esp given time limits.

File: README.md
Comment (brikis98) [reply README.md:204]: Ah, saw your message on Slack. Ignore this.

Comment (pete0emerson) [reply README.md:204]: Yes, the implementation plan for all to enjoy. Certainly will be removed at the end!

File: README.md
Comment (pete0emerson) [reply README.md:82]: Perfect, sounds good to me.

Comment (pete0emerson) [reply README.md:82]: One area worth touching on is `--tag`, should `--ref` be able to do the tag filtering that `--tag` is doing? I'm not quite sure yet how to do that if so, since `--ref` is more generic. Perhaps leave the `--tag` examples alone?

Comment (brikis98) [reply README.md]: IIRC, tags are valid refs—given a ref `xxx`, Git will try it as a tag and a branch or something like that, and you can of course force it to be a tag via `/refs/tags/xxx`—but prob worth testing how Git handles those on the CLI!

Comment (pete0emerson) [reply README.md]: Agreed. I'm thinking about how it might handle `--ref=>1.0.5`. I haven't looked at that bit of the code too much yet.

Comment (brikis98) [reply README.md]: Oh, I gotcha, the version constraints. Yea, those prob need to be parsed before passing on to Git.

**Comment by pete0emerson** [comment]
This closes #73

File: README.md
Comment (pete0emerson) [reply README.md]: I've put this first and also solved the handling of tag constraints.

**Review by bwhaley — COMMENTED** [review]
Overall this LGTM! The docs are clear, the tests cases seem sufficient, and the implementation looks clean. I'm not a CODEOWNER on this repo and I haven't contributed to this project before, so I'll defer the approval to one of the CODEOWNERs.

My only open question though is this: it seems like`--ref` supersedes `--tag`, `--branch`, and `--commit`, so why keep those flags around?

**Comment by pete0emerson** [comment]
> My only open question though is this: it seems like`--ref` supersedes `--tag`, `--branch`, and `--commit`, so why keep those flags around?

The only reason for keeping the other flags around is for backwards compatibility. Once any uses of it are moved over to `--ref`, those could be trashed. I'd do that in a separate PR.

**Comment by ina-stoyanova** [comment]
Looks good & easy to understand!

File: main.go
Comment (ina-stoyanova) [inline main.go:164]: do you always need to provide one of these two? - e.g. a ref or a tagConstraint

**Review by ina-stoyanova — APPROVED** [review]
LGTM 🎉

File: main.go
Comment (pete0emerson) [reply main.go:164]: Okay, had to re-familiarize myself with the code a bit to get you what I believe is the right answer here.

You can now specify --ref instead of --commit, --branch, or --tag. If you specify ref, then we detect whether it might be a tag constraint and capture that. If you specify a tag, then we detect that constraint and capture that. It's confusing, but I think it becomes less confusing if/when we deprecate --commit, --branch, and --tag and just have everyone use --ref.

I can't yet comment on the timeline on such a deprecation, since I don't know who is relying on `fetch` and what automation tools might need to change by doing such a deprecation.

Comment (pete0emerson) [reply main.go:164]: Resolving this, but if you'd like more discussion about it, reopen.

**Pull request merged**
